staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,297 @@
+}
+
+func (m *Manager) WaitForStorageVersionUpdate(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition) error {
+	var latestSVUpdateInfo *storageVersionUpdateInfo
+	val, found := latestStorageVersionUpdateInfo.Load(crd.UID)
+	if !found {
+		return nil
+	}
+
+	latestSVUpdateInfo = val.(*storageVersionUpdateInfo)
+
+	// NOTE: currently the graceful CRD deletion waits 1s for in-flight requests
+	// to register themselves to the wait group. Ideally the storage version update should
+	// not cause the requests to miss the 1s window; otherwise the requests may
+	// fail ungracefully (e.g. it may happen if the CRD was deleted immediately after the
+	// first CR request establishes the underlying storage).
+	select {
+	case <-latestSVUpdateInfo.updateChannels.errCh:
+		return fmt.Errorf("error while waiting for CRD storage version update")
+	case <-latestSVUpdateInfo.updateChannels.processedCh:
+		return nil
+	case <-ctx.Done():
+		return fmt.Errorf("aborted waiting for CRD storage version update: %w", ctx.Err())
+	// Unblock the requests if the storage version update takes a long time, otherwise
+	// CR requests may stack up and overwhelm the API server.
+	// TODO(roycaihw): benchmark the storage version update latency to adjust the timeout.
+	case <-time.After(storageVersionUpdateTimeout):
+		return fmt.Errorf("timeout waiting for CRD storage version update")
+	}
+
+}
+
+func (m *Manager) RunUpdateLoop(stopCh <-chan struct{}) {
+	go wait.Until(m.worker, time.Second, stopCh)
+	defer m.queue.ShutDown()
+	<-stopCh
+}
+
+func (m *Manager) worker() {
+	for m.processLatestSVUpdate() {
+	}
+}
+
+func (m *Manager) processLatestSVUpdate() bool {
+	ctx := context.TODO()
+	key, quit := m.queue.Get()
+	if quit {
+		return false
+	}
+	defer m.queue.Done(key)
+
+	val, ok := latestStorageVersionUpdateInfo.Load(key)
+	latestSVUpdateInfo := val.(*storageVersionUpdateInfo)
+	if !ok || latestSVUpdateInfo.processed {
+		klog.V(2).Infof("No pending storageversion update found, returning")
+		return true
+	}
+
+	m.updateStorageVersion(ctx, latestSVUpdateInfo.crd, latestSVUpdateInfo.updateChannels.teardownFinishedCh, latestSVUpdateInfo.updateChannels.processedCh, latestSVUpdateInfo.updateChannels.errCh)
+	markUpdateAsProcessed(latestSVUpdateInfo)
+	return true
+}
+
+func (m *Manager) recordLatestSVUpdateInfoWithTeardown(crd *apiextensionsv1.CustomResourceDefinition, tearDownFinishedCh <-chan struct{}) {
+	svUpdateInfo := &storageVersionUpdateInfo{
+		crd: crd,
+		updateChannels: &storageVersionUpdateChannels{
+			processedCh:        make(chan struct{}),
+			errCh:              make(chan struct{}),
+			teardownFinishedCh: tearDownFinishedCh,
+		},
+	}
+
+	// overwrite existing SVUpdateInfo with latest update event.
+	m.updateLatestStorageVersionUpdateInfo(crd.UID, svUpdateInfo)
+}
+
+func (m *Manager) updateLatestStorageVersionUpdateInfo(crdUID types.UID, svUpdateInfo *storageVersionUpdateInfo) {
+	latestStorageVersionUpdateInfo.Store(crdUID, svUpdateInfo)
```

Comment:
should entries be deleted from this map? It looks like is only ever grows.

Reply:
Yeah that part is still pending. Entries are made per CRD and should be deleted only when a CRD is deleted I think. Any other cases where you think we can delete the entry?

Reply:
Added functionality to delete entries from the map.